provision/docker/container/container.go

```diff
@@ -387,48 +387,56 @@ func (c *Container) Exec(client provision.BuilderDockerClient, stdout, stderr io
 
 // Commits commits the container, creating an image in Docker. It then returns
 // the image identifier for usage in future container creation.
-func (c *Container) Commit(client provision.BuilderDockerClient, limiter provision.ActionLimiter, writer io.Writer) (string, error) {
+func (c *Container) Commit(client provision.BuilderDockerClient, limiter provision.ActionLimiter, writer io.Writer, isDeploy bool) (string, error) {
 	log.Debugf("committing container %s", c.ID)
 	parts := strings.Split(c.BuildingImage, ":")
 	if len(parts) < 2 {
 		return "", log.WrapError(errors.Errorf("error parsing image name, not enough parts: %s", c.BuildingImage))
 	}
 	repository := strings.Join(parts[:len(parts)-1], ":")
 	tag := parts[len(parts)-1]
-	opts := docker.CommitContainerOptions{Container: c.ID, Repository: repository, Tag: tag}
-	done := limiter.Start(c.HostAddr)
-	image, err := client.CommitContainer(opts)
-	done()
-	if err != nil {
-		return "", log.WrapError(errors.Wrapf(err, "error in commit container %s", c.ID))
-	}
-	imgHistory, err := client.ImageHistory(c.BuildingImage)
-	imgSize := ""
-	if err == nil && len(imgHistory) > 0 {
-		fullSize := imgHistory[0].Size
-		if len(imgHistory) > 1 && strings.Contains(imgHistory[1].CreatedBy, "tail -f /dev/null") {
-			fullSize += imgHistory[1].Size
+	tags := []string{tag}
+	if isDeploy && tag != "latest" {
+		tags = append(tags, "latest")
+	}
+	for i, tag := range tags {
+		opts := docker.CommitContainerOptions{Container: c.ID, Repository: repository, Tag: tag}
+		done := limiter.Start(c.HostAddr)
+		image, err := client.CommitContainer(opts)
```

Comment:
Calling `client.CommitContainer` for each tag is not necessary and it may cause a performance hit, it should be called only once (for the original tag), after the Commit is done we should call `client.TagImage` to tag the latest tag. This will also allow us to move the commit logic and the image size logic to outside the for loop (removing the ugly `if i == 0` block).
Only the `PushImage` logic should be inside this for loop.